core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -2666,8 +2648,8 @@ class ReplicaManager(val config: KafkaConfig,
           //   is unavailable. This is required to ensure that we include the partition's
           //   high watermark in the checkpoint file (see KAFKA-1647).
           val state = info.partition.toLeaderAndIsrPartitionState(tp, isNew)
-          val isNewLeaderEpoch = partition.makeFollower(state, offsetCheckpoints, Some(info.topicId))
-
+          val isNewLeaderEpoch = partition.makeFollower(state, offsetCheckpoints, Some(info.topicId), partitionDirectoryId)
+          maybeUpdateTopicAssignment(partition, partitionDirectoryId)
```

Comment:
I think we should only call this if `localChanges.leaders` or `localChanges.followers` aren't empty, otherwise we can get duplicate assignment requests.

Reply:
this already happening as `applyLocalFollowersDelta` and `applyLocalLeadersDelta`  are only invoked `ReplicaManager.applyDelta` when `localChanges.followers` or `localChanges.leaders` aren't empty.  https://github.com/apache/kafka/blob/4325f9c6a3c3458485f0f7f3e718a63d63e72087/core/src/main/scala/kafka/server/ReplicaManager.scala#L2574

Reply:
Apologies, you're right. I still think does can create duplicates.

I think maybe we should extend `TopicDelta.localChanges` to also consider changes to the assigned directory to replicas hosted in the broker. Then we can gate calling `maybeUpdateTopicAssignment` unless `localChanges.directoryIds` is empty.